larq_compute_engine/core/indirect_bgemm/kernel.h

```diff
@@ -1,123 +1,195 @@
+    const std::int32_t dilation_height = bconv2d_params->dilation_height_factor;
+    const std::int32_t dilation_width = bconv2d_params->dilation_width_factor;
+    const std::int32_t input_padding_top =
+        bconv2d_params->padding_values.height;
+    const std::int32_t input_padding_left =
+        bconv2d_params->padding_values.width;
+    const std::int32_t batch_size = bitpacked_input_shape.Dims(0);
+    const std::int32_t input_height = bitpacked_input_shape.Dims(1);
+    const std::int32_t input_width = bitpacked_input_shape.Dims(2);
+    const std::int32_t output_height = output_shape.Dims(1);
+    const std::int32_t output_width = output_shape.Dims(2);
+    const std::int32_t output_size = num_output_pixels;
+    const std::int32_t tiled_output_size =
+        block_size_pixels *
+        ((output_size + block_size_pixels - 1) / block_size_pixels);
+
+    // Create the indirection buffer with padding (+ block_size_pixels) and fill
+    // it with pointers to the first element of the input, so that the padding
+    // at the end of the array contains pointers to valid memory.
+    indirection_buffer.assign(
+        tiled_output_size * filter_size + block_size_pixels, input_ptr);
+    // Assign the zero buffer that will be used for padding.
+    zero_buffer.assign(filter_size * input_depth, 0);
+
+    for (std::int32_t output_tile_start = 0;
+         output_tile_start < tiled_output_size;
+         output_tile_start += block_size_pixels) {
+      for (std::int32_t output_tile_offset = 0;
+           output_tile_offset < block_size_pixels; output_tile_offset++) {
+        const std::int32_t output_index =
+            std::min(output_tile_start + output_tile_offset, output_size - 1);
+        const std::int32_t batch_index =
+            output_index / (output_height * output_width);
+        const std::int32_t output_x = output_index % output_width;
+        const std::int32_t output_y =
+            (output_index % (output_height * output_width)) / output_width;
+        for (std::int32_t f_y = 0; f_y < filter_height; f_y++) {
+          const std::int32_t input_y = output_y * stride_height +
+                                       f_y * dilation_height -
+                                       input_padding_top;
+          if (0 <= input_y && input_y < input_height) {
+            for (std::int32_t f_x = 0; f_x < filter_width; f_x++) {
+              const std::int32_t input_x = output_x * stride_width +
+                                           f_x * dilation_width -
+                                           input_padding_left;
+              const std::int32_t kernel_index = f_y * filter_width + f_x;
+              const std::int32_t index = output_tile_start * filter_size +
+                                         kernel_index * block_size_pixels +
+                                         output_tile_offset;
+              if (0 <= input_x && input_x < input_width) {
```

Comment:
Note sure how speed-critical this function is, but if it is this if-statement will hurt a bit I guess. We could consider separating the inner-loop(s) into a part without an if-statement and a part with for the border parts.

Reply:
I think this is only called at the beginning of the kernel, so it should be fine. But we could also use the bound check utility function from our internal `types.h` which checks this condition with a single comparison:
```cpp
// Check that 0 <= index < limit using a single comparison, assuming
// that 0 <= limit if Index is signed.  Intended for use in performance
// critical contexts where 0 <= index < limit is almost always true.
inline bool FastBoundsCheck(const int index, const int limit) {
  return LCE_LIKELY((unsigned)index < (unsigned)limit);
}
```